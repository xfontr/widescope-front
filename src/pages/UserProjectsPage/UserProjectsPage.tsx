import { useEffect } from "react";
import { useAppSelector } from "../../app/hooks";
import Projects from "../../components/Projects/Projects";
import useProjects from "../../hooks/useProjects/useProjects";

const UserProjectsPage = (): JSX.Element => {
  const { getByAuthor } = useProjects();
  const { id, isLogged } = useAppSelector((state) => ({
    id: state.user.user.id,
    isLogged: state.user.isLogged,
  }));

  useEffect(() => {
    (async () => {
      isLogged && (await getByAuthor(id!));
    })();
  }, [getByAuthor, id, isLogged]);

  const projects = useAppSelector((state) =>
    isLogged ? state.userData.projects : []
  );

  return (
    <>
      <h2 className="page__title">
        Your creations.
        <span className="page__title--bold"> Thanks for sharing!</span>
      </h2>

      {projects.length ? (
        <Projects projects={projects} isReadOnly={false} />
      ) : (
        <span>
          You don't seem to have any project up. What about sharing one of your
          works?
        </span>
      )}
    </>
  );
};

export default UserProjectsPage;
