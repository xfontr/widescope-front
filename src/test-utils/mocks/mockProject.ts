import { IProject } from "../../types/project";

const mockProject: IProject = {
  id: "630e31ecb968115ba6a3e29f",
  name: "Project123",
  description: "The description",
  technologies: ["React", "Express"],
  repository: "www.google.com",
  author: "username",
  authorId: "id",
  logo: "ProjectLogo.png",
  logoBackup: "BackupProjectLogo.png",
  creationDate: "1662627402346",
};

export default mockProject;
