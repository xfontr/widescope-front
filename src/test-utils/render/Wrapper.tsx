import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { store } from "../../app/store";
import GlobalStyles from "../../styles/GlobalStyles";
import styledMainTheme from "../../styles/styledMainTheme";

interface WrapperProps {
  children: JSX.Element | JSX.Element[];
}

const Wrapper = ({ children }: WrapperProps): JSX.Element => {
  return (
    <ThemeProvider theme={styledMainTheme}>
      <Provider store={store}>
        <BrowserRouter>
          <GlobalStyles />
          {children}
        </BrowserRouter>
      </Provider>
    </ThemeProvider>
  );
};

export default Wrapper;
