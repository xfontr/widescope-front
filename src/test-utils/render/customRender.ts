import { render } from "@testing-library/react";
import { Wrapper } from "./Wrapper";

const customRender = (view: JSX.Element, ...options: any) =>
  render(view, { wrapper: Wrapper, ...options });

export * from "@testing-library/react";

export { customRender as render };
