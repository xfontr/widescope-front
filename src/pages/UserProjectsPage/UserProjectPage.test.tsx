import { render as reactRender } from "@testing-library/react";
import mockProject from "../../test-utils/mocks/mockProject";
import { render, screen } from "../../test-utils/render/customRender";
import { WrapperWithMockStore } from "../../test-utils/render/Wrapper";
import UserProjectsPage from "./UserProjectsPage";

describe("Given a UserProjectsPage component", () => {
  describe("When instantiated", () => {
    test("Then it should show a title and the projects retreived", async () => {
      reactRender(<UserProjectsPage />, { wrapper: WrapperWithMockStore });

      const heading = screen.getByRole("heading", {
        name: "Your creations. Thanks for sharing!",
      });

      const cards = await screen.findAllByRole("heading", {
        name: mockProject.name,
        level: 3,
      });

      expect(heading).toBeInTheDocument();
      expect(cards).toHaveLength(1);
    });
  });

  describe("When intantiated but no projects are found or the user is not logged in", () => {
    test("Then it should show a message 'No projects found.'", async () => {
      render(<UserProjectsPage />);

      const cards = screen.queryByRole("heading", {
        name: mockProject.name,
        level: 3,
      });

      const message = await screen.findByText(
        "You don't seem to have any project up. What about sharing one of your works?"
      );

      expect(cards).toBeNull();
      expect(message).toBeInTheDocument();
    });
  });
});
