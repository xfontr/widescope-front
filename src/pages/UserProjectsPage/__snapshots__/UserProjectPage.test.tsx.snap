// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a UserProjectsPage component When instantiated Then it should match the snapshot 1`] = `
Array [
  <h2
    className="page__title"
  >
    Your creations.
    <span
      className="page__title--bold"
    >
       Thanks for sharing!
    </span>
  </h2>,
  <span>
    You don't seem to have any project up. What about sharing one of your works?
  </span>,
]
`;
